react/src/components/ComputeSessionNodeItems/ContainerLogModal.tsx

```diff
@@ -0,0 +1,269 @@
+import _ from 'lodash';
+import { DownloadIcon } from 'lucide-react';
+import React, { useEffect, useRef, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { useFragment } from 'react-relay';
+
+interface ContainerLogModalProps extends BAIModalProps {
+  sessionFrgmt: ContainerLogModalFragment$key;
+}
+
+const ContainerLogModal: React.FC<ContainerLogModalProps> = ({
+  sessionFrgmt,
+  ...modalProps
+}) => {
+  const baiClient = useSuspendedBackendaiClient();
+  const { token } = theme.useToken();
+
+  const session = useFragment(
+    graphql`
+      fragment ContainerLogModalFragment on ComputeSessionNode {
+        id
+        row_id @required(action: NONE)
+        name
+        status
+        access_key
+        kernel_nodes {
+          edges {
+            node {
+              id
+              row_id
+              container_id
+              cluster_role
+            }
+          }
+        }
+      }
+    `,
+    sessionFrgmt,
+  );
+
+  const kernelNodes = session?.kernel_nodes?.edges?.map((e) => e?.node) || [];
+  const [selectedKernelId, setSelectedKernelId] = useState(
+    _.find(kernelNodes, (e) => e?.cluster_role === 'main')?.row_id ||
+      kernelNodes[0]?.row_id,
+  );
+
+  // Currently we can only fetch full logs
+  // const [logSize, setLogSize] = useState<100|'full'>('full');
+
+  const {
+    data: logs,
+    refetch,
+    isPending,
+    isRefetching,
+    dataUpdatedAt,
+  } = useTanQuery<string>({
+    queryKey: [
+      'containerLog',
+      session?.row_id,
+      session?.access_key,
+      selectedKernelId,
+      modalProps.open,
+    ],
+    queryFn: async () => {
+      if (
+        !modalProps.open ||
+        !session?.row_id ||
+        !session?.access_key ||
+        !selectedKernelId
+      ) {
+        return '';
+      }
+      return await baiClient
+        .get_logs(session?.row_id, session?.access_key, selectedKernelId, 15000)
+        .then((req: any) => req.result.logs);
+    },
+    staleTime: 5000,
+  });
+
+  // let queryParams: Array<string> = [];
+  // if (session?.access_key != null) {
+  //   queryParams.push(`owner_access_key=${session.access_key}`);
+  // }
+  // if (baiClient.supports('per-kernel-logs') && selectedKernelId !== null) {
+  //   queryParams.push(`kernel_id=${selectedKernelId}`);
+  // }
+  // let queryString = `/session/${session?.row_id}/logs`;
+  // if (queryParams.length > 0) {
+  //   queryString += `?${queryParams.join('&')}`;
+  // }
+  // // const url = `${baiClient._endpoint}${queryString}`
+
+  // const signed = baiClient.newSignedRequest('GET', queryString, null);
+  // console.log(signed)
+  // console.log(signed.uri);
+
+  const previousLastLineNumber = useRef(0);
+
+  useEffect(() => {
+    previousLastLineNumber.current = logs?.split('\n').length || 0;
+  }, [logs]);
+
+  const { md } = Grid.useBreakpoint();
+  const { t } = useTranslation();
+  const fixedPreviousLastLineNumber = previousLastLineNumber.current;
```

Comment:
Is there a reason why you wrapped the ref once here?

Reply:
`previousLastLineNumber.current` is updated in the above `useEffect` and is used in the `ScrollFollow`'s render props. Without storing it like this, the "previous last line number" used in `ScrollFollow` may already be the value updated by the `useEffect`. This is because `ScrollFollow` can render at any time, regardless of `ContainerLogModal` rendering.